torch/csrc/jit/ir.h

```diff
@@ -70,6 +70,65 @@ struct SourceLocation {
   std::string python_traceback;
 };
 
+// Scope is a node of a trie that represents the tree of nested scopes.
+// Individual scopes are pushed and popped from Graph, which holds a
+// pointer to the current scope. Each Node in Graph holds a pointer
+// to the scope that was current when the node was created.
+// The trie never needs to shrink, it only grows until it is disposed
+// of when Graph is deallocated. Hence, pointers to scopes held by nodes
+// will always be valid as long as Graph is alive.
+struct Scope {
+private:
+  Scope* parent_;
+  Symbol name_;
+  std::unordered_set<std::unique_ptr<Scope> > children_;
+public:
+  Scope() {
+    name_ = stringToSymbol("");
+    parent_ = NULL;
+  }
+  Scope(Scope* parent, Symbol name) {
+    name_ = name;
+    parent_ = parent;
+  }
+  Scope* push(Symbol name) {
+    Scope* newScope = new Scope(this, name);
+    children_.insert(std::unique_ptr<Scope>(newScope));
+    return newScope;
```

Comment:
Using set is fine, but as far as I understand the main purpose is to make it a true `trie`, but this is not really what happens in here - you will return the new unbound scope even if the insertion failed, which in turn will cause the pointer to be dangling, because the `unique_ptr` will be destructed immediately.

Now I'm wondering if we really want to keep it as a true trie... Maybe it just makes sense to have a new scope for each invocation of a module on a particular level (so you can actually see a sequence of scopes - if you call A B A, then you will see A B A in the scope tree).

 Maybe let's really use a vector and skip deduplication.

Reply:
Uh oh, sorry about this one. In fact, I was thinking about it the wrong way when I said you would re-enter a scope.
The insert would always succeed here, since the scope is always a new scope and the set will by construction not contain it. Which is equivalent to using a vector and appending to it, so I'm just going to use a vector.

Deduping based on names only makes sense if we'll ever compare the `scope_` pointers stored in Node to check if two nodes have the same scope. I'm going without, let me know about this.